core/dbt/context/providers.py

```diff
@@ -1410,7 +1410,8 @@ def sql(self) -> Optional[str]:
         if self.model.language == ModelLanguage.sql:  # type: ignore[union-attr]
             # If the model is deferred and the adapter doesn't support zero-copy cloning, then select * from the prod
             # relation
-            if getattr(self.model, "defer_relation", None):
+            # TODO: avoid routing on args.which if possible
+            if getattr(self.model, "defer_relation", None) and self.config.args.which == "clone":
```

Comment:
out of scope for this PR, but one way to do this could be to have the clone task further mutate the manifest after `merge_from_artifact` is called to overwrite `compiled_code` based on `defer_relation`. that way we could get rid of this if block entirely